src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1954,201 @@ MutableBigInteger sqrt() {
         }
     }
 
+    /**
+     * Calculate the integer square root {@code floor(sqrt(this))} and the remainder
+     * if needed, where {@code sqrt(.)} denotes the mathematical square root.
+     * The contents of {@code this} are <em>not</em> changed.
+     * The value of {@code this} is assumed to be non-negative.
+     *
+     * @return the integer square root of {@code this} and the remainder if needed
+     * @throws ArithmeticException if the value returned by {@code bitLength()}
+     * overflows the range of {@code int}.
+     */
+    MutableBigInteger[] sqrtRem(boolean needRemainder) {
+        // Special cases.
+        if (this.intLen <= 2) {
+            final long x = this.toLong(); // unsigned
+
+            /* For every long value s in [0, 2^32) such that x == s * s,
+             * it is true that s == Math.round(Math.sqrt(x >= 0 ? x : x + 0x1p64)),
```

Comment:
To reinforce this claim, I suggest to add a test for all perfect squares.

Reply:
I did it, although I'm afraid it takes up too much running time due to the overhead of BigInteger's wrapping...

Reply:
I mean only restricted to unsigned `long` perfect squares, something like the following, but written as a proper test

```
        long i = 0;
        for (; i < 1L << 32; ++i) {
            long x = i * i;
            long s = (long) Math.sqrt(x >= 0 ? x : x + 0x1p64);
            if (!(s + 1 == i || s == i)) {
                System.out.format("oops... i=%d, but s=%d%n", i, s);
                System.exit(1);
            }
        }
```

Reply:
It takes about 5 s on my laptop.

Reply:
> I mean only restricted to unsigned `long` perfect squares, something like the following, but written as a proper test
> 
> ```
>         long i = 0;
>         for (; i < 1L << 32; ++i) {
>             long x = i * i;
>             long s = (long) Math.sqrt(x >= 0 ? x : x + 0x1p64);
>             if (!(s + 1 == i || s == i)) {
>                 System.out.format("oops... i=%d, but s=%d%n", i, s);
>                 System.exit(1);
>             }
>         }
> ```

So, it is correct although the code does not test directly BigInteger.sqrt()?

Reply:
I think there's a misunderstanding here.

What I'd like to see is a test that verifies the claim
```
        /* For every long value s in [0, 2^32) such that x == s * s,
         * it is true that s - 1 <= (long) Math.sqrt(x >= 0 ? x : x + 0x1p64) <= s,
```
A test similar to my code above should be more than enough.

Reply:
That test would not exercise any code in `MutableBigInteger` or `BigInteger`. It would just ensure that `MBI.ulongSqrt()` has solid foundations.

Feel free to add the test at your discretion. However, if you decide for it, please make sure that it describes in comments what its purpose really is.